feat: implement recursive placeholder resolution in Default parameters

AshokThangavel opened the pull request:
## **Support Recursive Placeholder Resolution in Module Parameter**

### **Description**
This PR introduces deeply nested variables or chained configuration settings where one `${variable}` depends on another.
 with a safety handle complex dependencies and prevent infinite loops in the case of circular references.

fixes: #1009

### **Changes**

* **Recursive Logic:** Updated `ResolvePlaceholders` to use a `while` loop that continues as long as substitutions are being made.
* **Efficiency:** The loop exits immediately if a pass results in no changes, ensuring no performance penalty for simple configurations.


### **Testing Performed**

1. **Unit Tests:** Verified with complex nested strings:
* `${A}` -> `${B}` -> `${C}` -> `FinalValue`.

2. **Integration Test:** Added a new integration test class `Test.PM.Integration.Module` that:
* Generates a `module.xml` from XData.
* Writes it to the filesystem.
* Executes an IPM `load` command to verify the full lifecycle of the variable resolution.
### Test Execution Results: `Test.PM.Integration.Module`
| Counter | Action | Status | Description |
| --- | --- | --- | --- |
| 1 | LogMessage | passed | start Loading the demo-module1 module |
| 2 | AssertStatusOK | passed | Directory created /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 3 | AssertStatusOK | passed | module.xml File created on /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 4 | AssertStatusOK | passed | Created the xml file on /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 5 | AssertStatusOK | passed | Loaded varresolver module successfully from /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 6 | AssertTrue | passed | Module demo-module1 exists in IPM and version is 1.0.0 |
| 7 | LogMessage | passed | List all modules |
| 8 | AssertStatusOK | passed | uninstalled module demo-module1 successfully. |
| 9 | AssertStatusOK | passed | Deleted the module.xml file from /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 10 | LogMessage | passed | Duration of execution: 1.193466 sec. |

AshokThangavel committed:
feat: implement recursive placeholder resolution in Default parameters

AshokThangavel committed:
fix: update method name

AshokThangavel committed:
docs: update the CHANGELOG file

isc-dchui commented on CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - #822: The CPF resource processor now supports system expressions and macros in CPF merge files
 - #578 Added functionality to record and display IPM history of install, uninstall, load, and update
 - #961: Adding creation of a lock file for a module by using the `-create-lockfile` flag on install.
+- #1013: Implement recursive placeholder resolution in Default parameters
```
Move this to the new 0.10.6 section please

isc-dchui reviewed (changes requested):
A few comments and questions

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

AshokThangavel committed:
fix: updated the code based on feedback

AshokThangavel commented on tests/integration_tests/Test/PM/Integration/Module.cls:
```diff
@@ -0,0 +1,88 @@
+			<Default Name="ipmtest" Value="TESTING MY STRING"/>
+			<Default Name="ipmdir" Value="/usr/irissys/mgr/user/mts"/>
+			<Default Name="datapath" Value="${libdir}data/" />
+			<Default Name="datapath1" Value="${ipmdir}data/" />
```
Thank you for pointing out that excellent scenario. You are absolutely right—this is a 'pathological' but critical edge case for configuration flexibility. I have updated the implementation.

AshokThangavel committed:
refactor: Updated placeholder logic
-Support nested ${var} and {$var} delimiters and handled Frankenstein-case

AshokThangavel commented on src/cls/IPM/Storage/ModuleSetting/Default.cls:
```diff
@@ -57,6 +57,40 @@ ClassMethod EvaluateAttribute(
+            continue:data'["${"
+
+            set varExpr = "${" _ $piece($piece(data, "${", 2), "}") _ "}"
+            set resolved = ##class(%IPM.Utils.Module).%EvaluateSystemExpression(varExpr)
```
I've modified the logic to ensure the system variables and custom variables.

AshokThangavel commented on tests/integration_tests/Test/PM/Integration/Module.cls:
```diff
@@ -0,0 +1,88 @@
+    do $$$AssertTrue($isobject(module), "Module "_..#ModuleName_" exists in IPM and version is "_ module.Version.ToString())
+
+        do $$$LogMessage("List all modules")
+    set status = ##class(%IPM.Main).Shell("list")
```
Thank you!. I've added the `$$$AssertStatusOK`

AshokThangavel commented on src/cls/IPM/Storage/ModuleSetting/Default.cls:
```diff
@@ -57,6 +57,40 @@ ClassMethod EvaluateAttribute(
     return attribute
 }
 
+ClassMethod ResolvePlaceholders(ByRef customParams)
```
Thank you. Added the comments on top of method `ResolvePlaceholders`

AshokThangavel committed:
fix:updated the code based on feedback

isc-dchui reviewed (commented):
Thanks for the updates! Here's another pass of comments/questions

AshokThangavel committed:
refactor: implement robust variable resolution using %EvaluateSystemExpression
- Adopted %EvaluateSystemExpression for system variable resolution to align with existing APIs.
- Retained multi-pass logic to handle nested and "Frankenstein" placeholders.
- Implemented a priority gate: customParams take precedence over system defaults.
- Hardened maxLevels condition using an explicit guard clause for better robustness.

isc-dchui reviewed (commented):
Just a few more things, thanks!

AshokThangavel committed:
fix: code updated based on comments

AshokThangavel committed:
Merge remote-tracking branch 'origin/main' into feat/unified-variable-interpolation

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

AshokThangavel committed:
docs: updated changelog

AshokThangavel commented:
I’ve pulled the latest changes from main and resolved the conflicts in CHANGELOG.md. The new entry has been updated to `0.10.7`. Thank you!

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

AshokThangavel committed:
Merge branch 'feat/unified-variable-interpolation' of https://github.com/AshokThangavel/ipm into feat/unified-variable-interpolation

AshokThangavel committed:
Merge branch 'main' into feat/unified-variable-interpolation

isc-dchui committed:
Merge branch 'main' into feat/unified-variable-interpolation

isc-dchui committed:
Merge branch 'main' into feat/unified-variable-interpolation
